internal/scrapertest/compare.go

```diff
@@ -122,14 +133,20 @@ func CompareNumberDataPointSlices(expected, actual pdata.NumberDataPointSlice) e
 		}
 	}
 
-	matchingDPS := make(map[pdata.NumberDataPoint]pdata.NumberDataPoint, actual.Len())
-	for j := 0; j < actual.Len(); j++ {
+	once = make(map[int]bool, numPoints)
+
+	matchingDPS := make(map[pdata.NumberDataPoint]pdata.NumberDataPoint, numPoints)
+	for j := 0; j < numPoints; j++ {
```

Comment:
Do we still need the second loop now? Can we just report gaps in `once` map from the previous walk?

Reply:
Gaps in `once` can tell us which data points were missing, but doesn't tell us which data points are extra. This is enough to say that the two slices are not equal, but I think it's worth it to have clear errors, to quickly troubleshoot issues.

Maybe there is a more concise or performant way to do both, but I'm not seeing it yet.

Reply:
Maybe I don't fully understand the logic here. For me it looks like the first loop fully covers both cases now.

We iterate over all the expected datapoints slice, and report all missing datapoints, but at the same time we fill `once` map which is something like map[actualSliceIndex]isSeen, right? So any `actual.At(idx)` when `once[idx] == false` should be an extra datapoint. And we can replace the second iteration with something like this:

```
for j := 0; j < numPoints; j++ {
  if !once[j] {
    errs = multierr.Append(errs, fmt.Errorf("metric has extra datapoint with attributes: %v", actual.At(j).Attributes().AsRaw()))
  }
}
```

Or am I missing something?

Reply:
I was missing it, thanks for clarifying. 

There was a map being populated by the second loop, but moving this up to the first loop is even better, as it can serve the same purpose as `once`.